fvm/src/kernel/default.rs

```diff
@@ -265,36 +282,51 @@ where
     C: CallManager,
 {
     fn block_open(&mut self, cid: &Cid) -> Result<(BlockId, BlockStat)> {
-        // TODO(M2): Check for reachability here.
-
         self.call_manager
             .charge_gas(self.call_manager.price_list().on_block_open_base())?;
 
         let start = GasTimer::start();
 
+        if !self.blocks.is_reachable(cid) {
+            return Err(syscall_error!(NotFound; "block not reachable: {cid}").into());
+        }
+
         let data = self
             .call_manager
             .blockstore()
             .get(cid)
             // TODO: This is really "super fatal". It means we failed to store state, and should
             // probably abort the entire block.
             .or_fatal()?
-            .ok_or_else(|| anyhow!("missing state: {}", cid))
+            .ok_or_else(|| anyhow!("missing reachable state: {}", cid))
             // Missing state is a fatal error because it means we have a bug. Once we do
             // reachability checking (for user actors) we won't get here unless the block is known
             // to be in the state-tree.
             .or_fatal()?;
```

Comment:
This dance of checking `or_fatal()`, `ok_or_else()` and then `or_fatal()` again seems a bit redundant to me

Reply:
Yeah, fair enough. I've collapsed it.